File: vertx-web/src/main/resources/vertx-web-js/template_engine.js
/*
 * Copyright 2014 Red Hat, Inc.
 *
 * Red Hat licenses this file to you under the Apache License, version 2.0
 * (the "License"); you may not use this file except in compliance with the
 * License.  You may obtain a copy of the License at:
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

/** @module vertx-web-js/template_engine */
var utils = require('vertx-js/util/utils');
var Buffer = require('vertx-js/buffer');
var RoutingContext = require('vertx-web-js/routing_context');

var io = Packages.io;
var JsonObject = io.vertx.core.json.JsonObject;
var JTemplateEngine = io.vertx.ext.web.templ.TemplateEngine;

/**
 A template engine uses a specific template and the data in a routing context to render a resource into a buffer.
 <p>
 Concrete implementations exist for several well-known template engines.

 @class
*/
var TemplateEngine = function(j_val) {

  var j_templateEngine = j_val;
  var that = this;

  /**
   Render

   @public
   @param context {RoutingContext} the routing context 
   @param templateFileName {string} the template file name to use 
   @param handler {function} the handler that will be called with a result containing the buffer or a failure. 
   */
  this.render = function(context, templateFileName, handler) {
    var __args = arguments;
    if (__args.length === 3 && typeof __args[0] === 'object' && __args[0]._jdel && typeof __args[1] === 'string' && typeof __args[2] === 'function') {
      j_templateEngine["render(io.vertx.ext.web.RoutingContext,java.lang.String,io.vertx.core.Handler)"](context._jdel, templateFileName, function(ar) {
      if (ar.succeeded()) {
        handler(utils.convReturnVertxGen(ar.result(), Buffer), null);
      } else {
        handler(null, ar.cause());
      }
    });
    } else throw new TypeError('function invoked with invalid arguments');
  };

  // A reference to the underlying Java delegate
  // NOTE! This is an internal API and must not be used in user code.
  // If you rely on this property your code is likely to break if we change it / remove it without warning.
  this._jdel = j_templateEngine;
};

// We export the Constructor function
module.exports = TemplateEngine;

File: vertx-web-common/src/main/java/io/vertx/ext/web/common/template/TemplateEngine.java
/*
 * Copyright 2014 Red Hat, Inc.
 *
 *  All rights reserved. This program and the accompanying materials
 *  are made available under the terms of the Eclipse Public License v1.0
 *  and Apache License v2.0 which accompanies this distribution.
 *
 *  The Eclipse Public License is available at
 *  http://www.eclipse.org/legal/epl-v10.html
 *
 *  The Apache License v2.0 is available at
 *  http://www.opensource.org/licenses/apache2.0.php
 *
 *  You may elect to redistribute this code under either of these licenses.
 */

package io.vertx.ext.web.common.template;

import io.vertx.codegen.annotations.GenIgnore;
import io.vertx.codegen.annotations.VertxGen;
import io.vertx.core.Future;
import io.vertx.core.buffer.Buffer;
import io.vertx.core.json.JsonObject;

import java.util.Map;

/**
 * A template template uses a specific template and the data in a routing context to render a resource into a buffer.
 * <p>
 * Concrete implementations exist for several well-known template engines.
 *
 * @author <a href="http://tfox.org">Tim Fox</a>
 */
@VertxGen
public interface TemplateEngine {

  /**
   * Render the template. Template engines that support partials/fragments should extract the template base path from
   * the template filename up to the last file separator.
   *
   * Some engines support localization, for these engines, there is a predefined key "lang" to specify the language to
   * be used in the localization, the format should follow the standard locale formats e.g.: "en-gb", "pt-br", "en".
   *
   * @param context  the routing context
   * @param templateFileName  the template file name to use
   * @return a future notified with a result containing the buffer or a failure
   */
  default Future<Buffer> render(JsonObject context, String templateFileName) {
    return render(context.getMap(), templateFileName);
  }

  /**
   * Render the template. Template engines that support partials/fragments should extract the template base path from
   * the template filename up to the last file separator.
   *
   * Some engines support localization, for these engines, there is a predefined key "lang" to specify the language to
   * be used in the localization, the format should follow the standard locale formats e.g.: "en-gb", "pt-br", "en".
   *
   * @param context  the routing context
   * @param templateFileName  the template file name to use
   * @return a future notified with a result containing the buffer or a failure
   */
  @GenIgnore(GenIgnore.PERMITTED_TYPE)
  Future<Buffer> render(Map<String, Object> context, String templateFileName);

  /**
   * Returns the underlying engine, so further configurations or customizations may be applied or {@code null} when the
   * engine cannot unwrap it.
   * @return the engine instance.
   */
  @GenIgnore(GenIgnore.PERMITTED_TYPE)
  default Object unwrap() {
    return null;
  }

  /**
   * Clears any internal caches used by this engine. For most engines this means clearing the
   * vert.x cache, implementations that do not use vert.x as a cache should clear their own cache.
   */
  void clearCache();
}
